stl/inc/ranges

```diff
@@ -6821,6 +6821,682 @@ namespace ranges {
+
+        constexpr _Sentinel(_Sentinel<!_IsConst> _Rhs) noexcept(
+            is_nothrow_move_constructible_v<_Sentinel_tuple_type>) // strengthened
+            requires(_IsConst
+                     && (convertible_to<sentinel_t<_ViewTypes>, sentinel_t<_Maybe_const<_IsConst, _ViewTypes>>> && ...))
+            : _End(_STD move(_Rhs._End)) {}
+
+        template <bool _IteratorConst>
+            requires(sentinel_for<sentinel_t<_Maybe_const<_IsConst, _ViewTypes>>,
+                iterator_t<_Maybe_const<_IteratorConst, _ViewTypes>>>&&...)
+        _NODISCARD_FRIEND constexpr bool operator==(const _Iterator_type<_IteratorConst>& _Lhs,
+            const _Sentinel& _Rhs) noexcept(_Iterator_sentinel_equality_noexcept<_IteratorConst>) /* strengthened */ {
+            return _Zip::_Evaluate_iterator_sentinel_equality(_Get_iterator_tuple(_Lhs), _Rhs._End);
+        }
+
+        template <bool _IteratorConst>
+            requires(sized_sentinel_for<sentinel_t<_Maybe_const<_IsConst, _ViewTypes>>,
+                iterator_t<_Maybe_const<_IteratorConst, _ViewTypes>>>&&...)
+        _NODISCARD_FRIEND constexpr _Iterator_sentinel_difference_type<_IteratorConst>
+            operator-(const _Iterator_type<_IteratorConst>& _Lhs, const _Sentinel& _Rhs) noexcept(
+                noexcept(_Zip::_Get_smallest_iterator_sentinel_distance<
+                    common_type_t<range_difference_t<_Maybe_const<_IteratorConst, _ViewTypes>>...>>(
+                    _Get_iterator_tuple(_Lhs), _Rhs._End))) /* strengthened */
+        {
+            using _Difference_type = common_type_t<range_difference_t<_Maybe_const<_IteratorConst, _ViewTypes>>...>;
+            return _Zip::_Get_smallest_iterator_sentinel_distance<_Difference_type>(
+                _Get_iterator_tuple(_Lhs), _Rhs._End);
+        }
+
+        template <bool _IteratorConst>
+            requires(sized_sentinel_for<sentinel_t<_Maybe_const<_IsConst, _ViewTypes>>,
+                iterator_t<_Maybe_const<_IteratorConst, _ViewTypes>>>&&...)
+        _NODISCARD_FRIEND constexpr _Iterator_sentinel_difference_type<_IteratorConst> operator-(const _Sentinel& _Lhs,
+            const _Iterator_type<_IteratorConst>& _Rhs) noexcept(noexcept(-(_Rhs - _Lhs))) /* strengthened */ {
+            return -(_Rhs - _Lhs);
+        }
+
+    private:
+        template <bool _IteratorConst>
+        _NODISCARD static constexpr const auto& _Get_iterator_tuple(const _Iterator_type<_IteratorConst>& _Itr) {
+            return zip_view<_ViewTypes...>::_Get_iterator_tuple(_Itr);
+        }
+    };
+
+    namespace _Zip {
+        template <class _ResultType, class _LHSItrOrSentTuple, class _RHSItrOrSentTuple>
+            requires(tuple_size_v<decay_t<_LHSItrOrSentTuple>> == tuple_size_v<decay_t<_RHSItrOrSentTuple>>)
+        _NODISCARD constexpr _ResultType _Get_smallest_iterator_sentinel_distance(_LHSItrOrSentTuple&& _Lhs_tuple,

```

Comment:
This function seems unnecessarily complex. I believe we can do the exact same thing that `size()` does here, except that we need to check the first pair for negative or positive to tell us whether to call `min` or `max`.
The following should work:

```cxx
    template <size_t _FirstIdx /* = 0 */, size_t... _Idxs>
    _ResultType _Get_smallest_iterator_sentinel_distance(...) {
        _ResultType _First_size = _STD get<0>(_Lhs_tuple) - _STD get<0>(_Rhs_tuple);

        if (_First_size == 0) {
            return 0;
        }
        const initializer_list<_ResultType> _Sizes = {_First_size, (_STD get<_Idxs>(_Lhs_tuple) - _STD get<_Idxs>(_Rhs_tuple))...}; 
        if (_First_size < 0) {
            return (_STD max)(_Sizes);
        }
        if (_First_size > 0) {
            return (_STD min)(_Sizes)
        }
```

Reply:
Technically the distances between pairs of iterators might not have the same sign, but I believe that it only happens when an iterator is out of bound, and it's weird in general, so I think it makes sense to use `max` and `min`.

```c++
std::forward_list list = {1, 2, 3, 4, 5};
std::vector vec = {1, 2, 3, 4, 5};
std::counted_iterator cit{list.begin(), 5};

subrange sublist{cit, std::next(cit, 2)};

auto v = views::zip(sublist, vec);

auto it = std::next(v.begin(), 3);
auto end = v.end();

it - end; // 1 according to the standard; -2 with `min`
```

Reply:
yeah, this is a case where the requirements for iterators means that `max`/`min` must have the correct behavior here, and so under as-if we can do the simple `max`/`min`